File: rb/my_stack.rb
load 'my_node.rb'

class MyStack
  attr_accessor :head, :min_stack, :size

  def initialize
    @head = nil
    @min_stack = nil
    @size = 0
  end

  def push(node)
    node = MyNode.new(node)
    if @size.zero?
      @head = node
      @min_stack = node.dup
    elsif node.key < @min_stack.key # new node is less than min
      new_min = node.dup
      new_min.next_pointer = @min_stack # new min to last min node
      @min_stack = new_min # min node is now new node
    end
    node.next_pointer = @head if @size > 0
    @head = node
    @size += 1
  end

  def top
    size.zero? ? nil : @head.key
  end

  def pop
    return nil if @size.zero?

    current_node = @head
    if @size == 1
      @head = nil
      @min_stack = nil
    elsif @min_stack.key == @head.key
      @min_stack = @min_stack.next_pointer
    end
    @head = current_node.next_pointer
    @size -= 1
    current_node.key
  end

  def min
    return nil if @min_stack.nil?

    @min_stack.key
  end
end

File: js/my_stack.js
const my_node = require('./my_node');

class MyStack {
    constructor(head=null, min_stack=null, size=0) {
        this.head = head;
        this.min_stack = min_stack;
        this.size = size;
    };

    push(data) {
        const node = new my_node.MyNode(data);
        if (this.size === 0) {
            this.head = node;
            this.min_stack = {...node};
        } else if (node.key < this.min_stack.key) {
            let new_min = {...node};
            new_min.next_pointer = this.min_stack;
            this.min_stack = new_min;
        } else if (this.size > 0) {
            node.next_pointer = this.head;
        }
        this.head = node;
        this.size++;
    };

    pop() {
        if (this.size === 0) {return null};
        let current_node = this.head;
        if (this.size === 1) {
            this.head = null;
            this.min_stack = null;
            this.size = this.size - 1;
            return current_node;
        } else if (this.min_stack.key == this.head.key) {
            this.min_stack = this.min_stack.next_pointer;
        }
        this.head = current_node.next_pointer;
        this.size = this.size - 1;
        return current_node.key
    };

    top() {
        return this.head
    };

    min() {
        return this.min_stack.key
    };
}

module.exports = { MyStack }